# EXAMÉN CIENCIA DE DATOS PARA LA ECONOMÍA: IGNACIO CHANDIA #

In [ ]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
df = pd.read_csv('nba_logreg2.csv', sep = ";")
df.head(3)

A continuación se presenta una dataset sobre métricas de rendimientos de jugadores de la NBA durante su carrera, junto con la variable target si la carrera del basquetbolista supera o no los 5 años.

A continuación se detalla las variables que componen este dataset:

| Nombre columnas | Contenido |
|--- | --- |
|Name   |Nombre |
| GP    |Juegos jugados |
| MIN   |Minutos jugados |
| PTS   |Puntos por partido |
| FGM   |Objetivos de campo realizados |
| FGA   |Objetivos de campo intentados |
| FG%   |Porcentaje de puntos de campo |
| 3P Mode   |3 puntos hecho |
| 3PA   |Intentos de 3 puntos |
| 3P%   |3 puntos por ciento |
| FTM   |Tiro libre hecho |
| FTA   |Intentos de lanzamiento libre |
| FT%   |Porcentaje de lanzamiento libre |
| OREB  |Rebotes ofensivos |
| DREB  |Rebotes defensivos |
| REB   |Rebotes |
| AST   |Asistencias |
| STL   |Roba |
| BLK   |Bloques |
| TOV   |Pérdidas de balón |
| TARGET_5Yrs   |Variable predictora (1-Si la duración de la carrera>= 5 años) y (0-si<5) |



In [ ]:
# Descripción de los datos
print(df.describe())

In [ ]:
print(df.head())

In [ ]:
df.info()

In [ ]:
variablescontinuas = [
    'MIN',
    'PTS',
    'FGM',
    'FGA',
    'FG%',
    '3P Made',
    '3PA',
    '3P%',
    'FTM',
    'FTA',
    'FT%',
    'OREB',
    'DREB',
    'REB',
   'AST',
    'STL',
    'BLK',
    'TOV'
]

df_1 = df[variablescontinuas]

df_1.head()

En el paso anterior, se quitó la columna name, pues no es continua y genera problemas con las visualizaciones, ademas de la variable predecida.

In [ ]:
plt.figure(figsize=(16, 12))
correlation_matrix = df_1.corr()
sns.heatmap(correlation_matrix, annot=True, cmap='coolwarm', vmin=-1, vmax=1)
plt.title('Mapa de Calor de la Matriz de Correlación')
plt.show()

La primera parte del proyecto fue realizada con éxito, donde fueron descritas las variables y se visualizan con el gráfico de correlación, para ver la relación de cada variable.


# PARTE 3 LAZY


In [25]:
pip install lazypredict


Note: you may need to restart the kernel to use updated packages.


In [24]:
from lazypredict.Supervised import LazyClassifier
from sklearn.model_selection import train_test_split
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.preprocessing import StandardScaler

In [26]:
df = pd.read_csv('nba_logreg2.csv', sep = ";")
df.head(3)

Name  GP   MIN  PTS  FGM  FGA   FG%  3P Made  3PA   3P%  ...  \
0   Brandon Ingram  36 27.40 7.40 2.60 7.60 34.70     0.50 2.10 25.00  ...   
1  Andrew Harrison  35 26.90 7.20 2.00 6.70 29.60     0.70 2.80 23.50  ...   
2   JaKarr Sampson  74 15.30 5.20 2.00 4.70 42.20     0.40 1.70 24.40  ...   

   FTA   FT%  OREB  DREB  REB  AST  STL  BLK  TOV  TARGET_5Yrs  
0 2.30 69.90  0.70  3.40 4.10 1.90 0.40 0.40 1.30         0.00  
1 3.40 76.50  0.50  2.00 2.40 3.70 1.10 0.50 1.60         0.00  
2 1.30 67.00  0.50  1.70 2.20 1.00 0.50 0.30 1.00         0.00  

[3 rows x 21 columns]

In [27]:
df11 = df.drop('Name', axis=1)
df11

GP   MIN  PTS  FGM  FGA   FG%  3P Made  3PA   3P%  FTM  FTA   FT%  OREB  \
0     36 27.40 7.40 2.60 7.60 34.70     0.50 2.10 25.00 1.60 2.30 69.90  0.70   
1     35 26.90 7.20 2.00 6.70 29.60     0.70 2.80 23.50 2.60 3.40 76.50  0.50   
2     74 15.30 5.20 2.00 4.70 42.20     0.40 1.70 24.40 0.90 1.30 67.00  0.50   
3     58 11.60 5.70 2.30 5.50 42.60     0.10 0.50 22.60 0.90 1.30 68.90  1.00   
4     48 11.50 4.50 1.60 3.00 52.40     0.00 0.10  0.00 1.30 1.90 67.40  1.00   
...   ..   ...  ...  ...  ...   ...      ...  ...   ...  ...  ...   ...   ...   
1324  80 15.80 4.30 1.60 3.60 43.30     0.00 0.20 14.30 1.20 1.50 79.20  0.40   
1325  68 12.60 3.90 1.50 4.10 35.80     0.10 0.70 16.70 0.80 1.00 79.40  0.40   
1326  43 12.10 5.40 2.20 3.90 55.00     0.00 0.00  0.00 1.00 1.60 64.30  1.50   
1327  52 12.00 4.50 1.70 3.80 43.90     0.00 0.20 10.00 1.20 1.80 62.50  0.20   
1328  47 11.70 4.40 1.60 4.40 36.90     0.40 1.30 33.30 0.70 1.00 67.30  0.20   

      DREB  REB  AST  STL  BLK  TOV  TARGET_5Yrs  
0     3.40 4.10 1.90 0.40 0.40 1.30         0.00  
1     2.00 2.40 3.70 1.10 0.50 1.60         0.00  
2     1.70 2.20 1.00 0.50 0.30 1.00         0.00  
3     0.90 1.90 0.80 0.60 0.10 1.00         1.00  
4     1.50 2.50 0.30 0.30 0.40 0.80         1.00  
...    ...  ...  ...  ...  ...  ...          ...  
1324  0.80 1.20 2.50 0.60 0.20 0.80         0.00  
1325  1.10 1.50 2.30 0.80 0.00 1.30         1.00  
1326  2.30 3.80 0.30 0.30 0.40 0.90         0.00  
1327  0.40 0.70 2.20 0.40 0.10 0.80         1.00  
1328  0.70 0.90 1.40 0.70 0.10 0.90         1.00  

[1329 rows x 20 columns]

In [28]:
df11.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1329 entries, 0 to 1328
Data columns (total 20 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   GP           1329 non-null   int64  
 1   MIN          1329 non-null   float64
 2   PTS          1329 non-null   float64
 3   FGM          1329 non-null   float64
 4   FGA          1329 non-null   float64
 5   FG%          1329 non-null   float64
 6   3P Made      1329 non-null   float64
 7   3PA          1329 non-null   float64
 8   3P%          1329 non-null   float64
 9   FTM          1329 non-null   float64
 10  FTA          1329 non-null   float64
 11  FT%          1329 non-null   float64
 12  OREB         1329 non-null   float64
 13  DREB         1329 non-null   float64
 14  REB          1329 non-null   float64
 15  AST          1329 non-null   float64
 16  STL          1329 non-null   float64
 17  BLK          1329 non-null   float64
 18  TOV          1329 non-null   float64
 19  TARGET

In [29]:
y = df11['TARGET_5Yrs']
X = df11.drop('TARGET_5Yrs', axis=1) 

In [31]:
# Segmentación de datos

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=2023)


In [32]:
# Instanciar modelo
std = StandardScaler()
X_train_std = std.fit_transform(X_train)
X_test_std = std.fit_transform(X_test)

In [33]:
#Entrenado los modelos de regresión
models, predictions = LazyClassifier().fit(X_train_std, X_test_std, y_train, y_test)

100%|██████████| 29/29 [00:01<00:00, 22.14it/s]

[LightGBM] [Info] Number of positive: 657, number of negative: 406
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.000116 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 1723
[LightGBM] [Info] Number of data points in the train set: 1063, number of used features: 19
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.618062 -> initscore=0.481331
[LightGBM] [Info] Start training from score 0.481331


In [34]:
models[:10]

Accuracy  Balanced Accuracy  ROC AUC  F1 Score  \
Model                                                                         
SGDClassifier                    0.71               0.70     0.70      0.71   
BernoulliNB                      0.69               0.70     0.70      0.69   
PassiveAggressiveClassifier      0.69               0.69     0.69      0.70   
NearestCentroid                  0.67               0.69     0.69      0.67   
RidgeClassifierCV                0.73               0.68     0.68      0.72   
RidgeClassifier                  0.73               0.68     0.68      0.72   
LinearSVC                        0.72               0.68     0.68      0.71   
LogisticRegression               0.72               0.68     0.68      0.71   
CalibratedClassifierCV           0.72               0.68     0.68      0.71   
SVC                              0.73               0.68     0.68      0.71   

                             Time Taken  
Model                                    
SGDClassifier                      0.01  
BernoulliNB                        0.01  
PassiveAggressiveClassifier        0.01  
NearestCentroid                    0.01  
RidgeClassifierCV                  0.01  
RidgeClassifier                    0.01  
LinearSVC                          0.04  
LogisticRegression                 0.01  
CalibratedClassifierCV             0.15  
SVC                                0.04

### LazyClassifier

`LazyClassifier` es una biblioteca que permite entrenar múltiples modelos de clasificación de manera rápida y sencilla. Proporciona una comparación rápida de varios algoritmos sin necesidad de escribir mucho código. Esto es útil para identificar qué modelos de clasificación funcionan mejor con un conjunto de datos específico.

### Mejores Algoritmos de Clasificación

1. **Logistic Regression**: Utilizado para modelos de clasificación binaria. Es un método simple pero efectivo para problemas lineales.

2. **Random Forest**: Un conjunto de árboles de decisión entrenados en diferentes subconjuntos del conjunto de datos. Es robusto y reduce el riesgo de overfitting.

3. **Gradient Boosting**: Una técnica de machine learning para clasificación y regresión que construye un modelo de predicción fuerte a partir de un conjunto de modelos débiles.

4. **Support Vector Machines (SVM)**: Un algoritmo que encuentra el hiperplano que mejor separa las clases en el espacio de características.

5. **K-Nearest Neighbors (KNN)**: Un método simple que clasifica una muestra en función de la mayoría de los votos de sus k vecinos más cercanos.

Estos algoritmos fueron explicados y utilizados durante las clases, y cada uno tiene sus ventajas y desventajas dependiendo del tipo de datos y del problema específico a tratar.

Como vemos, todos los modelos son similares en cuanto a la precisión, pero el mejor modelo según las métricas generales (Accuracy, F1 Score y ROC AUC) parece ser RidgeClassifier y RidgeClassifierCV, pues ambos tienen:

Accuracy: 0.73 (el más alta)
F1 Score: 0.72 (el más alta entre los modelos de alta precisión)
